File: embedded/unittest/application_test/logger_test.rs
//! Unit tests for RAW and PCD application logger workers.

use std::{
    fs,
    path::PathBuf,
    sync::atomic::{AtomicU64, Ordering},
};

use crate::{models::pointcloud::PointXYZIRT, platform::pubsub::Topic};

use super::*;

static NEXT_FILE_ID: AtomicU64 = AtomicU64::new(0);

/// Creates a unique path in the operating system's temporary directory.
fn temporary_output(extension: &str) -> PathBuf {
    let id = NEXT_FILE_ID.fetch_add(1, Ordering::Relaxed);
    std::env::temp_dir().join(format!(
        "vista-lidar-logger-test-{}-{id}.{extension}",
        std::process::id()
    ))
}

/// Confirms that RAW logging preserves the exact sensor bytes.
#[test]
fn writes_raw_packets_without_modification() {
    let path = temporary_output("bin");
    let topic = Topic::new("test/lidar/raw", 2).unwrap();
    let subscriber = topic.subscribe("test-raw-logger").unwrap();
    let publisher = topic.publisher();

    publisher
        .publish(LidarMessage::new(
            "test-lidar",
            0,
            Some(10),
            20,
            RawPacket::new(vec![1, 2, 3, 4]),
        ))
        .unwrap();
    drop(publisher);

    let report = run_raw_logger(subscriber, path.clone()).unwrap();
    assert_eq!(report.message_count, 1);
    assert_eq!(fs::read(&path).unwrap(), vec![1, 2, 3, 4]);
    fs::remove_file(path).unwrap();
}

/// Confirms that PCD logging writes the final point count and point data.
#[test]
fn writes_processed_point_cloud_as_pcd() {
    let path = temporary_output("pcd");
    let topic = Topic::new("test/pointcloud/processed", 2).unwrap();
    let subscriber = topic.subscribe("test-pcd-logger").unwrap();
    let publisher = topic.publisher();
    let frame = PointCloudFrame::new(
        10,
        vec![PointXYZIRT {
            x: 1.0,
            y: 2.0,
            z: 3.0,
            intensity: 4,
            ring: 5,
            return_id: 0,
            timestamp_ns: 10,
        }],
    );

    publisher
        .publish(LidarMessage::new("test-lidar", 0, Some(10), 20, frame))
        .unwrap();
    drop(publisher);

    let report = run_pcd_logger(subscriber, path.clone()).unwrap();
    let contents = fs::read_to_string(&path).unwrap();
    assert_eq!(report.message_count, 1);
    assert_eq!(report.point_count, 1);
    assert!(contents.contains("POINTS 1"));
    assert!(contents.contains("1.000000 2.000000 3.000000 4 5 0"));
    fs::remove_file(path).unwrap();
}

File: embedded_C/unittest/application_test/logger_test.cpp
#include <atomic>
#include <filesystem>
#include <fstream>
#include <iterator>
#include <string>

#include "1_Platform/pubsub/pubsub.hpp"
#include "4_Applications/logger/logger.hpp"
#include "unittest/test.hpp"

namespace {

std::filesystem::path temporary_output(const std::string& extension) {
    static std::atomic<std::uint64_t> next_id{0};
    return std::filesystem::temp_directory_path() /
           ("vista-edge-cpp-test-" + std::to_string(next_id++) + "." + extension);
}

}  // namespace

VISTA_TEST(logger_preserves_raw_bytes) {
    const auto path = temporary_output("bin");
    vista::platform::Topic<vista::devices::LidarRawMessage> topic("test/raw", 2);
    auto subscriber = topic.subscribe("raw-logger");
    {
        auto publisher = topic.publisher();
        publisher.publish(vista::devices::LidarRawMessage(
            "test",
            0,
            10,
            20,
            vista::devices::RawPacket({1, 2, 3, 4})));
    }

    const auto report =
        vista::application::run_raw_logger(std::move(subscriber), path);
    std::ifstream input(path, std::ios::binary);
    const std::vector<std::uint8_t> bytes{
        std::istreambuf_iterator<char>(input),
        std::istreambuf_iterator<char>()};
    VISTA_CHECK(report.message_count == 1);
    VISTA_CHECK(bytes == std::vector<std::uint8_t>({1, 2, 3, 4}));
    input.close();
    std::filesystem::remove(path);
}

VISTA_TEST(logger_writes_pcd_header_and_point) {
    const auto path = temporary_output("pcd");
    vista::platform::Topic<vista::devices::LidarPointCloudMessage> topic(
        "test/pointcloud", 2);
    auto subscriber = topic.subscribe("pcd-logger");
    {
        auto publisher = topic.publisher();
        publisher.publish(vista::devices::LidarPointCloudMessage(
            "test",
            0,
            10,
            20,
            vista::models::PointCloudFrame(
                10,
                {vista::models::PointXYZIRT{1.0F, 2.0F, 3.0F, 4, 5, 0, 10}})));
    }

    const auto report =
        vista::application::run_pcd_logger(std::move(subscriber), path);
    std::ifstream input(path);
    const std::string contents{
        std::istreambuf_iterator<char>(input),
        std::istreambuf_iterator<char>()};
    VISTA_CHECK(report.message_count == 1);
    VISTA_CHECK(report.point_count == 1);
    VISTA_CHECK(contents.find("POINTS 1") != std::string::npos);
    VISTA_CHECK(
        contents.find("1.000000 2.000000 3.000000 4 5 0") != std::string::npos);
    input.close();
    std::filesystem::remove(path);
}
